Feat/alert/fetch profile picture of other users

Harrish92 opened the pull request:
# Fetch profile picture of other users

## Description
This PR introduces the functionality to fetch the profile pictures of other users in the Alert feature. It closes issue #324 .

#### Remarks :

- The `init` method is removed form `UserVIewModel` and re-written in `ProfileScreen.kt`. Its signature is changed to require a parameter of type `AuthenticationViewModel` . 
- The class signatures of `GPSServiceImpl` , `PushNotificationsServiceImpl`  are changed to require a parameter of type `AuthenticationViewModel` .
-  `_users` is added to `UserViewModel`  to represent a list of user data.

#### Steps to display profile pricture of other users:

1. Load data of other users using the `loadUsers` method from `UserViewModel` to get the `imageUrl` .
2. Use `imageUrl` as a parameter for `downloadFile` from `UserViewModel` to fetch the profile picture of other users.

## Changes
- Implemented a method to fetch all profile pictures from other users.
- Implemented a method to fetch user data from other users.
- Updated method to fetch profile picture of the user.
- Updated method to fetch user data of the authenticated user.
- Added error handling for cases where profile pictures cannot be fetched.
- Added unit tests handling the added methods.

## Files 

#### Added
- None

#### Modified
- `app/src/androidTest/java/com/android/periodpals/services/GPSServiceImplInstrumentedTest.kt`
- `app/src/main/java/com/android/periodpals/MainActivity.kt`
- `app/src/main/java/com/android/periodpals/model/user/UserModel.kt`
- `app/src/main/java/com/android/periodpals/model/user/UserModelSupabase.kt`
- `app/src/main/java/com/android/periodpals/model/user/UserViewModel.kt`
- `app/src/main/java/com/android/periodpals/services/GPSServiceImpl.kt`
- `app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt`
- `app/src/main/java/com/android/periodpals/ui/alert/CreateAlert.kt`
- `app/src/main/java/com/android/periodpals/ui/profile/ProfileScreen.kt`
- `app/src/test/java/com/android/periodpals/model/user/UserModelSupabaseTest.kt`
- `app/src/test/java/com/android/periodpals/model/user/UserViewModelTest.kt`
- `app/src/test/java/com/android/periodpals/services/GPSServiceImplTest.kt`
- `app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt`
- `app/src/test/java/com/android/periodpals/ui/profile/CreateProfileTest.kt`
- `app/src/test/java/com/android/periodpals/ui/profile/ProfileScreenTest.kt`
#### Removed
- None
## Dependencies Added
- None

## Testing
- Added Unit tests for the method fetching all profile pictures.
- Added Unit tests for the method fetching all user data.
- Mocked the authenticationViewModel.

taghizadlaura reviewed (commented):
## Summary
The code looks good overall, but it's unclear how we can directly verify that it works as intended (e.g., by visually displaying the fetched pictures). Should we wait for Ale's usage in her next PR, or is there another way to confirm that there are no bugs?

## Important
#### Code Quality
- The code is readable and consistent with the existing codebase.

#### Functionality
- While the implementation appears correct, there is uncertainty about how to ensure correctness without seeing it in action.

#### Performance
- No performance issues observed.

#### Security
- No security concerns noted.

## Testing
- The existing tests seem good; however, I still have the same question as mentioned in the summary section.

## Documentation
- While the documentation is generally well done, there are some missing parameters in various places. For example, the `onFailure` parameter for `upsertUserProfile()` should be included (See in comments).

## Suggestions
- Clarify how the implementation can be verified directly.
- Update the documentation to include all parameters for the relevant functions.

## Steps before PR approved
LGTM. Please address my question in the summary section and respond to my comments (e.g., add the missing parameters to the documentation).
I’ve left this review as a comment, but I’m available to approve it anytime; I’d just prefer to have an answer before it’s merged.

Harrish92 commented on app/build.gradle.kts:
I believe it is a merging problem

taghizadlaura reviewed (commented):
Thanks for your modifications! Could you please answer the question I raised in my previous review?

Harrish92 commented:
> Thanks for your modifications! Could you please answer the question I raised in my previous review?
(It's unclear how we can directly verify that it works as intended (e.g., by visually displaying the fetched pictures). Should we wait for Ale's usage in her next PR, or is there another way to confirm that there are no bugs?)

@taghizadlaura 
To answer to your question, let's look at the methods that were changed/added to the user model.

`loadUserProfile` : is modified to filter the user data to keep only the one from the authenticated user. The filter is similar to `deleteUserProfile`.

`loadUserProfiles` : is a copy of the old `loadUserProfile`, except that it receives now a list of user data. The RLS policy is also changed on Supabase to enable read access to all users. 

`downloadFilePublic`: uses the method `downloadPublic` which is a method from the Supabase Storage api to download files from other users.The RLS policy is also changed on Supabase to enable read access to all users.

Hope this answer helps !

taghizadlaura commented:
> @taghizadlaura To answer to your question, let's look at the methods that were changed/added to the user model.
> 
> `loadUserProfile` : is modified to filter the user data to keep only the one from the authenticated user. The filter is similar to `deleteUserProfile`.
> 
> `loadUserProfiles` : is a copy of the old `loadUserProfile`, except that it receives now a list of user data. The RLS policy is also changed on Supabase to enable read access to all users.
> 
> `downloadFilePublic`: uses the method `downloadPublic` which is a method from the Supabase Storage api to download files from other users.The RLS policy is also changed on Supabase to enable read access to all users.
> 
> Hope this answer helps !

Thanks for the clarification! Based on your suggestion, I can replace the parameter for the image URL with one that can be found in the `imageUrl` column of the `user` table on Supabase. However, I'm unsure if this is the best practice as it feels a bit complicated. That said, I will assume it's fine since the tests pass and the app doesn’t crash.

taghizadlaura reviewed (approved):
LGTM now that you’ve answered my question and addressed my comments.

agonzalez-r reviewed (approved):
## Summary
I had no way to verify that this actually works, but the code added to the user model/view model seems correct, very coherent with the rest, and you did the necessary changes in the supabase table, so it should work.
## Important (which includes)
#### Code Quality
- Overall, code is good but should be improved in some parts: I left some nitpicks in the comments, please go over them. 
- Be careful with the log messages of the added functions.
#### Functionality
- I'm not sure we need the old version of the functions (`loadUsers()` and `downloadFile()`), please explain why you think we need to keep both. Because loadUsers() for instance, is never used. 
#### Performance
- @taghizadlaura and I found a bug: when editing the profile, click on edit profile picture button (the little pencil), it goes to a "Select Photo" screen, from android, but when clicked on the back button, the app crashes. **This does not come from the new code**, the bug is also present in main.
#### Security 
Since you changed the policies in the user table in supabase to have the attributes **only readable**, it should be fine. 
## Testing
I guess you can't test the success of the `downloadFiles` function... but have you tried different mocking methods for the response of the repository? I know mocking the supabase response is very difficult in some cases, I did some heavy mocking for `getAlertsWithinRadius()` function in the alert model.

If it's really not possible, tests are as complete as they can be. 

## Documentation
Some minor documentation missing.

## Suggestions
- Fix the spotted bug in this PR, otherwise please create a separate task to do so.

## Steps before PR approved
- Explain why both version of the added functions to the user model/view model are needed, or remove one version if they aren't.
- Implement nitpicks I left in the comments.

Harrish92 commented on app/src/main/java/com/android/periodpals/model/user/UserModelSupabase.kt:
```diff
@@ -115,6 +133,23 @@ class UserRepositorySupabase(private val supabase: SupabaseClient) : UserReposit
+    }
+  }
+
   override suspend fun downloadFile(
```
You're remark is pertinent. Two methods are not necessary for download, since the bucket is public.

Harrish92 commented:
> I guess you can't test the success of the downloadFiles function... but have you tried different mocking methods for the response of the repository? I know mocking the supabase response id very difficult in some cases, I did some heavy mocking for getAlertsWithinRadius() function in the alert model.
>
>If it's really not possible, tests are as complete as they can be.

I tried to mock it in a previous PR, but my attempt did not prove to be fruitful.
